torch/nn/modules/module.py

```diff
@@ -596,76 +604,109 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
+        module in input :attr:`state_dict` is at ``state_dict._metadata[prefix]``.
+        Subclasses can achieve class-specific backward compatible loading using
+        the version number at ``state_dict._metadata[prefix]["version"]``.
 
-        When :attr:`is_missing` is ``True``, :attr:`full_name` can not be found in
-        the dict being loaded. When :attr:`is_missing` is ``False``,
-        :attr:`full_name` is unexpected in the dict being loaded.
+        .. note::
+            :attr:`state_dict` is not the same object as the input
+            :attr:`state_dict` to :meth:`~torch.nn.Module.load_state_dict`. So
+            it can be modified.
 
-        :attr:`name` is the actual name of the parameter/buffer, i.e., the
-        substring after the last `dot` in :attr:`full_name`.
+        Arguments:
+            state_dict (dict): a dict containing parameters and
+                persistent buffers.
+            prefix (str): the prefix for parameters and buffers used in this
+                module
+            strict (bool): whether to strictly enforce that the keys in
+                :attr:`state_dict` with :attr:`prefix` match the names of
+                parameters and buffers in this module
+            missing_keys (list of str): if ``strict=False``, add missing keys to
+                this list
+            unexpected_keys (list of str): if ``strict=False``, add unexpected
+                keys to this list
+            error_msgs (list of str): error messages should be added to this
+                list, and will be reported together in
+                :meth:`~torch.nn.Module.load_state_dict`
         """
-        pass
+        local_name_params = itertools.chain(self._parameters.items(), self._buffers.items())
+        local_state = {k: v.data for k, v in local_name_params if v is not None}
+
+        for name, param in local_state.items():
+            key = prefix + name
+            if key in state_dict:
+                input_param = state_dict[key]
+                if isinstance(input_param, Parameter):
+                    # backwards compatibility for serialized parameters
+                    input_param = input_param.data
+                try:
+                    param.copy_(input_param)
+                except Exception:
+                    error_msgs.append('While copying the parameter named "{}", '
+                                      'whose dimensions in the model are {} and '
+                                      'whose dimensions in the checkpoint are {}.'
+                                      .format(key, param.size(), input_param.size()))
+            elif strict:
+                missing_keys.append(key)
+
+        if strict:
+            for key, input_param in state_dict.items():
+                if key.startswith(prefix):
+                    input_name = key[len(prefix):]
+                    input_name = input_name.split('.', 1)[0]  # get the name of param/buffer/child
```

Comment:
Why is this necessary? Wouldn't it be enough to check that there's no `.` in `input_name`?

Reply:
This is necessary. If we just check for `.`, then we won't detect the submodules with names not in `self._modules`